File: src/tests/fixtures/analyzer/simple.ts
// Simple 3-round ping-pong implementation per specification
export interface SimpleMessage {
  type: "ping" | "pong";
  round: number;
}

class Channel<T> {
  private queue: T[] = [];
  private resolvers: ((value: T) => void)[] = [];

  send(value: T): void {
    if (this.resolvers.length > 0) {
      const resolve = this.resolvers.shift();
      if (resolve) {
        resolve(value);
      }
    } else {
      this.queue.push(value);
    }
  }

  receive(timeout = 5): Promise<T | null> {
    if (this.queue.length > 0) {
      const item = this.queue.shift();
      return Promise.resolve(item ?? null);
    }

    return new Promise((resolve) => {
      const timer = setTimeout(() => {
        const index = this.resolvers.indexOf(resolve);
        if (index !== -1) {
          this.resolvers.splice(index, 1);
        }
        resolve(null);
      }, timeout);

      this.resolvers.push((value: T) => {
        clearTimeout(timer);
        resolve(value);
      });
    });
  }
}

export async function simplePingPong(): Promise<SimpleMessage[]> {
  const messages: SimpleMessage[] = [];
  const pingChan = new Channel<SimpleMessage>();
  const pongChan = new Channel<SimpleMessage>();

  // Ping actor
  const pingActor = async () => {
    for (let round = 1; round <= 3; round++) {
      // Send ping
      const pingMsg: SimpleMessage = { type: "ping", round };
      console.log(`Ping: Sending round ${round}`);
      pongChan.send(pingMsg);
      messages.push(pingMsg);

      // Wait for pong
      const pongMsg = await pingChan.receive();
      if (pongMsg) {
        console.log(`Ping: Received pong ${pongMsg.round}`);
        messages.push(pongMsg);
      }
    }
  };

  // Pong actor
  const pongActor = async () => {
    for (let round = 1; round <= 3; round++) {
      // Receive ping
      const pingMsg = await pongChan.receive();
      if (pingMsg) {
        console.log(`Pong: Received ping ${pingMsg.round}`);

        // Send pong back
        const pongMsg: SimpleMessage = { type: "pong", round: pingMsg.round };
        pingChan.send(pongMsg);
        console.log(`Pong: Sent pong ${pongMsg.round}`);
      }
    }
  };

  // Run both actors concurrently
  await Promise.all([pingActor(), pongActor()]);

  return messages;
}

File: src/tests/fixtures/analyzer/simple.go
// Simple 3-round ping-pong implementation per specification
package main

import (
	"fmt"
	"sync"
	"time"
)

type SimpleMessage struct {
	Type  string
	Round int
}

func SimplePingPong(pingChan, pongChan chan SimpleMessage) []SimpleMessage {
	var messages []SimpleMessage
	var wg sync.WaitGroup
	messageMutex := &sync.Mutex{}

	// Ping actor
	wg.Add(1)
	go func() {
		defer wg.Done()
		for round := 1; round <= 3; round++ {
			// Send ping
			pingMsg := SimpleMessage{Type: "ping", Round: round}
			fmt.Printf("Ping: Sending round %d\n", round)
			pongChan <- pingMsg

			messageMutex.Lock()
			messages = append(messages, pingMsg)
			messageMutex.Unlock()

			// Wait for pong
			select {
			case pongMsg := <-pingChan:
				fmt.Printf("Ping: Received pong %d\n", pongMsg.Round)
				messageMutex.Lock()
				messages = append(messages, pongMsg)
				messageMutex.Unlock()
			case <-time.After(5 * time.Millisecond):
				fmt.Printf("Ping: Timeout waiting for pong\n")
			}
		}
	}()

	// Pong actor
	wg.Add(1)
	go func() {
		defer wg.Done()
		for round := 1; round <= 3; round++ {
			// Receive ping
			select {
			case pingMsg := <-pongChan:
				fmt.Printf("Pong: Received ping %d\n", pingMsg.Round)

				// Send pong back
				pongMsg := SimpleMessage{Type: "pong", Round: pingMsg.Round}
				pingChan <- pongMsg
				fmt.Printf("Pong: Sent pong %d\n", pongMsg.Round)
			case <-time.After(5 * time.Millisecond):
				fmt.Printf("Pong: Timeout waiting for ping\n")
			}
		}
	}()

	wg.Wait()
	return messages
}
